[MPS] Fix tensor with non-zero storage offset graph gathering

qqaatw opened the pull request:
Previously, the "can slice" flag in Placeholder constructor in `OperationUtils.mm` is conditioned on whether the numbers of dimensions of base shape and view shape are the same. This doesn't consider the situation that a view tensor could be the base tensor's sliced and then unsqueezed version, resulting in different num of dims.

For example, if we want to stack `y_mps` and `x_mps` on the last dim:
```
t_mps = torch.tensor([1, 2, 3, 4], device="mps")
x_mps = t_mps[2:]  # [3, 4]
y_mps = t_mps[:2]  # [1, 2]

res_mps = torch.stack((y_mps, x_mps), dim=-1)
```

the kernel will unsqueeze both of them on the last dim and then concatenate them, which is equivalent to:

```
res_mps = torch.cat((y_mps.unsqueeze(-1), x_mps.unsqueeze(-1)), dim=-1)
```

`x_mps.unsqueeze(-1)` is an unsqueezed and contiguous tensor with a storage offset, this kind of tensors should be sliceable without cloning its storage.

Fixes #87856
Fixes #91065


cc @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr @abhudev

qqaatw committed:
[MPS] Fix tensor with storage offset gathering

qqaatw committed:
Add test

qqaatw commented:
@pytorchbot label "module: mps"

malfet commented on test/test_mps.py:
```diff
@@ -1406,25 +1406,28 @@ def helper(operator):
                 res_mps = x_mps <= y_mps
                 res_cpu = x_cpu <= y_cpu
-            if operator == "<":
+            elif operator == "<":
```
This seems unrelated to change in question, is it?

qqaatw replied:
Yes not much related, but I think this improvement is trivial and small enough to not cause a reading difficulty. I feel it would be more costly and unnecessary to improve this in a separate PR, wouldn't it?

malfet commented on aten/src/ATen/native/mps/OperationUtils.mm:
```diff
@@ -226,7 +226,7 @@ void printTensorNDArray(const Tensor& t) {
   id<MTLBuffer> srcBuf = getMTLBufferStorage(src);
   // a view tensor could be contiguous (e.g., slice ops) or non-contiguous (e.g., transpose())
-  if (src.is_view() || !src.is_contiguous()) {
+  if (src.is_view() || src.storage_offset() != 0 || !src.is_contiguous()) {
```
Can you please add a more detailed description, why is this necessary? Shouldn't we just add handling for non-zero storage_offset to the MTLBuffer pointer computation?

qqaatw replied:
Let me find if there is a better way to handle this kind of tensor.

qqaatw replied:
For now I didn't find a proper way to get an offset of a MTLBuffer (from the [docs](https://developer.apple.com/documentation/metal/mtlbuffer?language=objc) on Appls's website there seems no such methods provided, not sure whether using primitive pointer operations to it is safe or not). What I know is that `MTLComputeCommandEncoder` derived from a `MTLCommandBuffer` in a custom metal kernel can add an offset for reading an existing buffer, like so `[computeEncoder setBuffer:a offset:0 atIndex:0];`, but it seems to not apply to this case. So unfortunately we may have to handle this as is.

Please let me know if I'm missing something.

kulinseth replied:
There is a way to add offset by creating Array view. You will have to tile the offset to all the dimensions as its done in the [code path ](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/mps/operations/View.mm#L488).

malfet reviewed (changes requested):
This PR contains unrelated changes as well as not much of an explanation as to why it is necessary to make a copy of the tensor view. 

I.e. it probably fixes the symptom, but not the fundamental underlying problem.

qqaatw committed:
wip

DenisVieriu97 committed:
[MPS] Fix bilinear backward pass (#94892)

Fixes backward pass for bilinear.

Summary of changes:
- bilinear op is able to produce **contiguous, non-view** tensors with a storage offset, such as: shape=`[1, 1, 1, 1]`, `storage_offset=12`. This seems a weird case, but it is valid, and for these type of tensors we wouldn't be able to gather/scatter since we look at the view flag (which is not set here). This change looks into `storage_offset` only rather than the is_view flag which is not being set
- **reduction sum** must return a zeroed out output if passing an input with 0 elements (e.g a shape of (0, 5)).
Pull Request resolved: https://github.com/pytorch/pytorch/pull/94892
Approved by: https://github.com/kulinseth

qqaatw committed:
wip

qqaatw committed:
cleanup

qqaatw committed:
Fix

qqaatw committed:
Fix lint

qqaatw commented:
@pytorchbot label "keep-going"

qqaatw committed:
Merge branch 'master' into fix_view_gather_logic

qqaatw committed:
wip

qqaatw committed:
Clean up

kulinseth commented on aten/src/ATen/native/mps/OperationUtils.mm:
```diff
@@ -289,7 +289,7 @@ void printTensorNDArray(const Tensor& t) {
     if (!mpsShape) {
       mpsShape = getMPSShape(_tensor);
-  }
+    }
```
Fix this.

qqaatw replied:
This is purposely an indentation fix for `if (!mpsShape) { ` block.

kulinseth commented:
Modulo the nit, changes look good.

qqaatw commented:
@pytorchbot merge

qqaatw commented:
@pytorchbot merge
